src/choker.cpp

```diff
@@ -168,13 +158,24 @@ namespace {
 		//   |             V             |
 		//   +---------------------------+
 		//   0%    num have pieces     100%
-		int const t1_total = t1->torrent_file().num_pieces();
-		int const t2_total = t2->torrent_file().num_pieces();
-		int const score1 = (lhs->num_have_pieces() < t1_total / 2
-			? t1_total - lhs->num_have_pieces() : lhs->num_have_pieces()) * 1000 / t1_total;
-		int const score2 = (rhs->num_have_pieces() < t2_total / 2
-			? t2_total - rhs->num_have_pieces() : rhs->num_have_pieces()) * 1000 / t2_total;
+		std::shared_ptr<torrent> const t = peer->associated_torrent().lock();
+		TORRENT_ASSERT(t);
+
+		std::int64_t const total_size = t->torrent_file().total_size();
+		std::int64_t have_size = peer->num_have_pieces() * t->torrent_file().piece_length();
```

Comment:
this looks like it would overflow. `num_have_pieces()` returns an `int`, right? and I imagine `piece_length()` does. This multiplication may overflow. You could either divide `uploaded_at_last_round()` by the piece length, or widen one of the factors before multiplying.

Also, I prefer the style of writing this as a single expression, so `have_size` can be made `const`.